src/Altcha.php

```diff
@@ -79,71 +72,108 @@ private static function decodePayload($payload)
         return $data;
     }
 
-    public static function createChallenge($options)
+    /**
+     * @param string|array<array-key, mixed> $data
+     */
+    private static function verifyAndBuildSolutionPayload($data): ?Payload
     {
-        if (is_array($options)) {
-            $options = new ChallengeOptions($options);
+        if (is_string($data)) {
+            $data = self::decodePayload($data);
         }
 
-        $algorithm = $options->algorithm ?: self::DEFAULT_ALGORITHM;
-        $maxNumber = $options->maxNumber ?: self::DEFAULT_MAX_NUMBER;
-        $saltLength = $options->saltLength ?: self::DEFAULT_SALT_LENGTH;
-
-        $params = $options->params;
-        if ($options->expires) {
-            $params['expires'] = $options->expires->getTimestamp();
+        if ($data === null
+            || !isset($data['algorithm'], $data['challenge'], $data['number'], $data['salt'], $data['signature'])
+            || !is_string($data['algorithm'])
+            || !is_string($data['challenge'])
+            || !is_int($data['number'])
+            || !is_string($data['salt'])
+            || !is_string($data['signature'])
+        ) {
+            return null;
         }
 
-        $salt = $options->salt ?: bin2hex(self::randomBytes($saltLength));
-        if (!empty($params)) {
-            $salt .= '?' . http_build_query($params);
+        return new Payload($data['algorithm'], $data['challenge'], $data['number'], $data['salt'], $data['signature']);
+    }
+
+    /**
+     * @param string|array<array-key, mixed> $data
+     */
+    private static function verifyAndBuildServerSignaturePayload($data): ?ServerSignaturePayload
+    {
+        if (is_string($data)) {
+            $data = self::decodePayload($data);
         }
 
-        $number = $options->number ?: self::randomInt($maxNumber);
+        if ($data === null
+            || !isset($data['algorithm'], $data['verificationData'], $data['signature'], $data['verified'])
+            || !is_string($data['algorithm'])
+            || !is_string($data['verificationData'])
+            || !is_string($data['signature'])
+            || !is_bool($data['verified'])
+        ) {
+            return null;
+        }
 
-        $challenge = self::hashHex($algorithm, $salt . $number);
+        return new ServerSignaturePayload($data['algorithm'], $data['verificationData'], $data['signature'], $data['verified']);
+    }
 
-        $signature = self::hmacHex($algorithm, $challenge, $options->hmacKey);
+    /**
+     * Creates a new challenge for ALTCHA.
+     *
+     * @param BaseChallengeOptions $options
+     *
+     * @return Challenge The challenge data to be passed to ALTCHA.
+     */
+    public static function createChallenge(BaseChallengeOptions $options): Challenge
```

Comment:
Intuitively, using a base class, `BaseChallengeOptions`, instead of an interface, e.g., `ChallengeOptionsInterface` doesn't make the code easier to read and still doesn't tell the user which class to instantiate. So I opted for a base class here.